webviz_subsurface/plugins/_group_tree/group_tree_data.py

```diff
@@ -395,42 +395,52 @@ def get_nodedict(gruptree: pd.DataFrame, nodename: str) -> Dict[str, Any]:
 def add_nodetype(gruptree: pd.DataFrame, smry: pd.DataFrame) -> pd.DataFrame:
     """Adds three columns to the gruptree dataframe: IS_PROD, IS_INJ and IS_OTHER.
 
-    Wells are classified as producers,  injectors or other based on the WSTAT summary data.
+    Wells are classified as producers, injectors or other based on the WSTAT summary data.
 
     Group nodes are classified as producing if it has any producing wells upstream and
-    correspondingly for injection and other.
+    correspondingly for injecting and other.
     """
-    df = pd.DataFrame()
+
+    ens_list = []
     for ensemble in gruptree["ENSEMBLE"].unique():
         gruptree_ens = gruptree[gruptree["ENSEMBLE"] == ensemble].copy()
         smry_ens = smry[smry["ENSEMBLE"] == ensemble].copy()
-        df_ens = add_nodetype_for_ens(gruptree_ens, smry_ens)
-        df = pd.concat([df, df_ens])
-    return df
+        ens_list.append(add_nodetype_for_ens(gruptree_ens, smry_ens))
+    return pd.concat(ens_list)
 
 
 def add_nodetype_for_ens(gruptree: pd.DataFrame, smry: pd.DataFrame) -> pd.DataFrame:
     """Adds nodetype IS_PROD, IS_INJ and IS_OTHER for an ensemble."""
 
+    # Get all nodes
     nodes = gruptree.drop_duplicates(subset=["CHILD"], keep="first").copy()
-    # Tag wells as producers, injector or other well
-    is_prod_map, is_inj_map, is_other_map = get_welltype_maps(
-        nodes[nodes["KEYWORD"] == "WELSPECS"], smry
+
+    # Identify leaf nodes (group nodes can also be leaf nodes)
+    def is_leafnode(node: pd.Series) -> bool:
+        if nodes[nodes["PARENT"] == node["CHILD"]].empty:
+            return True
+        return False
+
+    nodes["IS_LEAF"] = nodes.apply(is_leafnode, axis=1)
+
+    # Classify leaf nodes as producer, injector or other
+    is_prod_map, is_inj_map, is_other_map = create_leafnodetype_maps(
+        nodes[nodes["IS_LEAF"]], smry
     )
     nodes["IS_PROD"] = nodes["CHILD"].map(is_prod_map)
     nodes["IS_INJ"] = nodes["CHILD"].map(is_inj_map)
     nodes["IS_OTHER"] = nodes["CHILD"].map(is_other_map)
 
-    # Recursively find well types of all wells connected to the group node
+    # Recursively find well types of all leaf nodes connected to the group node
     # Deduce group node type from well types
-    groupnodes = nodes[nodes["KEYWORD"] != "WELSPECS"]
-    for _, groupnode in groupnodes.iterrows():
-        wells_are_prod, wells_are_inj, wells_are_other = get_leaf_well_types(
-            groupnode["CHILD"], nodes
+    nonleafs = nodes[nodes["IS_LEAF"].isin([False])]
```

Comment:
There are perhaps shorter notation for this? Does the "numpy" syntax also work for pandas? I.e.
```suggestion
    nonleafs = nodes[~nodes["IS_LEAF"]]
```

Reply:
Yes it works! I was searching for something like that, but I couldn't find it :)